From 5a613d627ca79bf99a8da974a508bc765abd14d0 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Mon, 3 Nov 2014 07:52:00 -0500
Subject: [infiniband] iser: Fix use after free in iser_snd_completion()

Message-id: <923a973c3c2bef4a2dd22139f0e04622852c7f65.1415000301.git.avadai@redhat.com>
Patchwork-id: 98671
O-Subject: [Patch RHEL7.1 02/37] IB/iser: Fix use after free in iser_snd_completion()
Bugzilla: 1107622
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Doug Ledford <dledford@redhat.com>

From: Dan Carpenter <dan.carpenter@oracle.com>

We use "tx_desc" again after we free it.

Signed-off-by: Dan Carpenter <dan.carpenter@oracle.com>
Acked-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit fd8b48b22a2b7cdf21f15b01cae379e6159a7eea)
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/ulp/iser/iser_initiator.c b/drivers/infiniband/ulp/iser/iser_initiator.c
index d257f19..0bd9833 100644
--- a/drivers/infiniband/ulp/iser/iser_initiator.c
+++ b/drivers/infiniband/ulp/iser/iser_initiator.c
@@ -610,11 +610,12 @@ void iser_snd_completion(struct iser_tx_desc *tx_desc,
   ib_dma_unmap_single(device->ib_device, tx_desc->dma_addr,
      ISER_HEADERS_LEN, DMA_TO_DEVICE);
   kmem_cache_free(ig.desc_cache, tx_desc);
+  tx_desc = NULL;
  }
 
  atomic_dec(&ib_conn->post_send_buf_count);
 
- if (tx_desc->type == ISCSI_TX_CONTROL) {
+ if (tx_desc && tx_desc->type == ISCSI_TX_CONTROL) {
   /* this arithmetic is legal by libiscsi dd_data allocation */
   task = (void *) ((long)(void *)tx_desc -
       sizeof(struct iscsi_task));
-- 
1.7.1